Ensure all needed options are added to association options list unconditionally

viralpraxis opened the pull request:
<!--
Thanks for contributing to Rails!

Please do not make *Draft* pull requests, as they still send
notifications to everyone watching the Rails repo.

Create a pull request when it is ready for review and feedback
from the Rails team :).

If your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

About this template

The following template aims to help contributors write a good description for their pull requests.
We'd like you to provide a description of the changes in your pull request (i.e. bugs fixed or features added), the motivation behind the changes, and complete the checklist below before opening a pull request.

Feel free to discard it if you need to (e.g. when you just fix a typo). -->

### Motivation / Background

<!--
Describe why this Pull Request needs to be merged. What bug have you fixed? What feature have you added? Why is it important?
If you are fixing a specific issue, include "Fixes #ISSUE" (replace with the issue number, remove the quotes) and the issue will be linked to this PR.
-->

This Pull Request has been created because is case of typo on `through` association option it's not listed in raised error's message.

For example, if we make a simple typo like this (t(h)rough)

```ruby
class User < ApplicationRecord
  has_many :courses
  has_many :assignments, trough: :courses
end
```

the error message on rails boot is kind of misleading:

>  Unknown key: :trough. Valid keys are: :class_name, :anonymous_class, :primary_key, :foreign_key, :dependent, :validate, :inverse_of, :strict_loading, :query_constraints, :autosave, :before_add, :after_add, :before_remove, :after_remove, :extend, :counter_cache, :join_table, :index_errors (ArgumentError)

There is not `through` which I expect to be noted.

### Detail

This PR fixes all of these issues for `HasOne` and `HasMany` association builders.

The reason of this is that `ActiveRecord::Associations::Builder::HasMany.valid_options` adds `through` only if it is present in provided options:

```ruby
def self.valid_options(options)
  ...
  valid += [:through, :source, :source_type] if options[:through]
  ...
end
```

That logic seems to be appropriate only for `source` and `source_type`, but I do not see why `through` can't be added in the list unconditionally. The same stands for `ActiveRecord::Associations::Builder::HasOne.valid_options`.

Applying patch fixes the problem:

> Unknown key: :trough. Valid keys are: :class_name, :anonymous_class, :primary_key, :foreign_key, :dependent, :validate, :inverse_of, :strict_loading, :query_constraints, :autosave, :before_add, :after_add, :before_remove, :after_remove, :extend, :counter_cache, :join_table, :index_errors, :through (ArgumentError)

### Additional information

If I didn't miss anything and there's an actual problem, there might be some other occurrences of it.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

skipkayhil commented:
The change makes sense to me :+1:

Can you add a test to ensure this doesn't regress? Additionally, I think most of the test failures are because your branch is out of date so I would rebase to fix those.

viralpraxis commented:
@skipkayhil thanks for the feedback. I rebased and added test. Note that there seems to be many of these issues, for example,  `:as` option behaves the same.

viralpraxis committed:
Ensure all association options are added to the list unconditionally

Some association options, e.g. `through`, were only added to
`.valid_options` list only if they were provided. It means that
sometimes ArgumentError's message would not mention all the
possible options.

viralpraxis commented:
I fixed all occurrences of described issue I could find for `HasOne` and `HasMany`. Also added a test to ensure all the "top-level" (i.e. which are not dependent on others, like `disable_joins` depends on `through`) options are noted in raised exception
